packages/vite/src/node/ssr/ssrExternal.ts

```diff
@@ -86,8 +88,97 @@ const CJS_CONTENT_RE =
 // TODO: use import()
 const _require = createRequire(import.meta.url)
 
-// do we need to do this ahead of time or could we do it lazily?
-function collectExternals(
+const isSsrExternalCache = new WeakMap<
+  ResolvedConfig,
+  (id: string) => boolean | undefined
+>()
+
+export function shouldExternalizeForSSR(
+  id: string,
+  config: ResolvedConfig
+): boolean | undefined {
+  let isSsrExternal = isSsrExternalCache.get(config)
+  if (!isSsrExternal) {
+    isSsrExternal = createIsSsrExternal(config)
+    isSsrExternalCache.set(config, isSsrExternal)
+  }
+  return isSsrExternal(id)
+}
+
+function createIsSsrExternal(
+  config: ResolvedConfig
+): (id: string) => boolean | undefined {
+  const processedIds = new Map<string, boolean | undefined>()
+
+  const { ssr, root } = config
+
+  const noExternal = ssr?.noExternal
+  const noExternalFilter =
+    noExternal !== 'undefined' &&
+    typeof noExternal !== 'boolean' &&
+    createFilter(undefined, noExternal, { resolve: false })
+
+  const isConfiguredAsExternal = (id: string) => {
+    const { ssr } = config
+    if (!ssr || ssr.external?.includes(id)) {
+      return true
+    }
+    if (typeof noExternal === 'boolean') {
+      return !noExternal
+    }
+    if (noExternalFilter) {
+      return noExternalFilter(id)
+    }
+    return true
+  }
+
+  const resolveOptions: InternalResolveOptions = {
+    root,
+    preserveSymlinks: config.resolve.preserveSymlinks,
+    isProduction: false,
+    isBuild: true
+  }
+
+  const isPackageEntry = (id: string) => {
+    if (!bareImportRE.test(id) || id.includes('\0')) {
+      return false
+    }
+    if (
+      tryNodeResolve(
+        id,
+        undefined,
+        resolveOptions,
+        ssr?.target === 'webworker',
+        undefined,
+        true
+      )
+    ) {
+      return true
+    }
+    try {
+      // no main entry, but deep imports may be allowed
+      const pkgPath = resolveFrom(`${id}/package.json`, root)
+      if (pkgPath.includes('node_modules')) {
+        return true
+      }
+    } catch {}
+    return false
+  }
+
+  return (id: string) => {
+    if (processedIds.has(id)) {
+      return processedIds.get(id)
+    }
+    const external =
+      isBuiltin(id) || (isPackageEntry(id) && isConfiguredAsExternal(id))
```

Comment:
maybe swap the order of these since `isPackageEntry` is probably more expensive so that it's sometimes short-circuited?
```suggestion
      isBuiltin(id) || (isConfiguredAsExternal(id) && isPackageEntry(id))
```

Reply:
Yes, probably a good idea. I did it the other way around because I thought `noExternal` could be a user-defined function at one point and better to avoid passing it things that we know aren't external. But we can check for that case if it ever happens 👍🏼